### PR TITLE
fix: Update Trial Download Error Message

### PR DESCRIPTION
**Pull request by MikhailKardash**
## Description

`Trial.list_checkpoints` never passed the `states` parameter, resulting in a return of STATE_DELETED checkpoints under some conditions. This is not an issue with the SDK, but rather a confusing output due to the CLI args. Updating the error message to dispel confusion.

## Test Plan

1. Find or make a trial with multiple checkpoints 
2. Get the UUID of the “best checkpoint”: det trial describe <trial_id> --json
3. Delete the checkpoint: det checkpoint delete <uuid>
4. Download the “best checkpoint”: det trial download <trial_id> --best

Expected behavior:  Another checkpoint is downloaded

## Checklist

- [X] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
MLG-1163

### CODE REVIEW

File: harness/determined/common/experimental/trial.py [hunk 1]
```diff
@@ -237,6 +237,7 @@ def get_trial_checkpoints(offset: int) -> bindings.v1GetTrialCheckpointsResponse
                 sortByMetric=sort_by if isinstance(sort_by, str) else None,
                 offset=offset,
                 limit=max_results,
+                states=[bindings.checkpointv1State.COMPLETED],
```
Comment (wes-turner) [inline harness/determined/common/experimental/trial.py:240]: Passing on a thread from OOB communication:

>         Only checkpoints in a ``COMPLETED`` state with a matching ``COMPLETED`` validation are considered.

Wonder what this means?

Comment (azhou-determined) [reply harness/determined/common/experimental/trial.py:240]: not quite sure, but i think maybe at some point we were only accepting sort parameters that were searcher metrics, which would imply that the checkpoint had to have a searcher validation completed. 

but since we allow sorting by various attributes (state, reported time, batch number) that aren't related to validation, i don't think this comment applies any more. 

so i think we should delete the comment. what's the motivation for this change though? i'm wondering if we should accept a `states` parameter to be passed in to `list_checkpoints`

Comment (MikhailKardash) [reply harness/determined/common/experimental/trial.py:240]: We can have `list_checkpoints` accept state, I like that more. Would this apply to both `harness/determined/common/experimental/trial.py:list_checkpoints` and `harness/determined/common/experimental/experiment.py:list_checkpoints` ?

Comment (wes-turner) [reply harness/determined/common/experimental/trial.py:240]: Now that Michael's refreshed my memory...
> what's the motivation for this change though?

Current behavior in the CLI's `det checkpoint download` can attempt to download a checkpoint that's been deleted, and `Checkpoint.download` is [explicitly opinionated](https://github.com/determined-ai/determined/blob/8924996c129f605c6779b3b952f689ee93b73cca/harness/determined/common/experimental/checkpoint/_checkpoint.py#L209-L217) about not downloading deleted checkpoints. From https://hpe-aiatscale.atlassian.net/browse/DET-3786, looks like maybe deleted checkpoints really are deleted? Then https://hpe-aiatscale.atlassian.net/browse/DET-9491 added "partial deletion" as a feature.

Anyway. So the bug is in the CLI. I'm also not sure changing `list_checkpoints` is the right place to fix it.

I tend to not like making our interface more complicated by adding options for things that can be done in user code unless necessary. Adding a filter like `state` is one of those things.

> So i think we should delete the comment.

I like this solution to making `list_checkpoints` match its docstring, too.

Comment (azhou-determined) [reply harness/determined/common/experimental/trial.py:240]: @wes-turner has a good point that adding a parameter to `list_checkpoints` has downstream implications. like it seems weird if we allow you to fetch a DELETED checkpoint but you can't actually do anything with it in the SDK. 

i do think adding `states` is a good idea, but it should probably require more thought. so my vote is we just fix the bug for now, instead of adding a new feature.

Comment (MikhailKardash) [reply harness/determined/common/experimental/trial.py:240]: So we can merge this fix now, and discuss SDK design in a follow-up? @azhou-determined @wes-turner

File: harness/determined/cli/trial.py [hunk 2]
```diff
@@ -181,7 +182,17 @@ def download(args: Namespace) -> None:
...
+    try:
+        path = checkpoint.download(path=args.output_dir)
+    except errors.CheckpointStateException:
+        print(
+            f"Invalid Checkpoint State for checkpoint {checkpoint.uuid}.\n",
+            "It's likely that this checkpoint is deleted. Please retry with a valid checkpoint UUID.",
+            sep=''
+        )
+        return
+    except Exception as e:
```
Comment (azhou-determined) [inline harness/determined/cli/trial.py:194]: why is this needed? unchecked exceptions (exceptions that are not `CheckpointStateException`) should be uncaught and raised automatically?

File: harness/determined/cli/trial.py [hunk 3]
```diff
@@ -181,7 +181,17 @@ def download(args: Namespace) -> None:
             raise ValueError(f"No checkpoints found for trial {args.trial_id}")
         checkpoint = checkpoints[0]
 
-    path = checkpoint.download(path=args.output_dir)
+    try:
+        path = checkpoint.download(path=args.output_dir)
+    except errors.CheckpointStateException:
+        print(
+            f"Invalid Checkpoint State for checkpoint {checkpoint.uuid}.",
```
Comment (azhou-determined) [inline harness/determined/cli/trial.py:188]: print statement is weird. do we need newlines? you should be able to just do 
```
print(
        f"Invalid Checkpoint State for checkpoint {checkpoint.uuid}. "
        "It's checkpoint has likely been deleted. Please retry with a valid checkpoint UUID."
    )
```

File: harness/determined/cli/trial.py (hunk 2 above)
Comment (MikhailKardash) [reply harness/determined/cli/trial.py]: Removed

File: harness/determined/cli/trial.py (hunk 3 above)
Comment (MikhailKardash) [reply harness/determined/cli/trial.py]: Changed it along with showing users checkpoint state.

File: harness/determined/cli/trial.py [hunk 4]
```diff
@@ -181,7 +181,15 @@ def download(args: Namespace) -> None:
             raise ValueError(f"No checkpoints found for trial {args.trial_id}")
         checkpoint = checkpoints[0]
 
-    path = checkpoint.download(path=args.output_dir)
+    try:
```
Comment (wes-turner) [inline harness/determined/cli/trial.py:184]: Printing an error message instead of letting Python render the uncaught exception with a stack trace is helpful. But I think the messaging is still is a little more confusing for the user than it could be

```
% det trial download 3
Invalid Checkpoint State for checkpoint 3ac811f1-d045-4d23-b573-612e853ccbe1. This checkpoint is in state CheckpointState.DELETED. Please retry with a different checkpoint UUID.
```

The person who typed `det trial download` didn't select a UUID, themselves. They _can't_ retry with a different checkpoint UUID because they didn't specify one in the first place.

I think the messaging most helpful here explains what happened during selection and what, if anything, the `det trial download` user can do to fix the problem. If the messaging can explain _why_ things ended up in this state, so much the better. Anything to give the user a sense of control and make it seem less like our system is buggy.

File: harness/determined/cli/trial.py [hunk 5]
```diff
@@ -181,7 +181,15 @@ def download(args: Namespace) -> None:
...
+        path = checkpoint.download(path=args.output_dir)
+    except errors.CheckpointStateException:
+        print(
+            f"Invalid Checkpoint State for checkpoint {checkpoint.uuid}. "
+            f"This checkpoint is in state {checkpoint.state}. "
+            "Please retry with a different checkpoint UUID."
+        )
+        return
```
Comment (wes-turner) [inline harness/determined/cli/trial.py:192]: Might also be worth looking at [how the CLI errors out](https://github.com/determined-ai/determined/blob/e9a457d75a021fd326bf2d970a0fe4f8a9062070/harness/determined/cli/cli.py#L291) in other situations and adopting that pattern. I dunno -- it's been a while since I've looked at it. Just a suggestion.

File: harness/determined/cli/trial.py (hunk 4 above)
Comment (MikhailKardash) [reply harness/determined/cli/trial.py:184]: I'm not sure I understand, I think all of the information is right there for the user. The checkpoint returned by `--best` is in STATE_DELETED. And to fix it, they can chose a checkpoint with a different UUID.

File: harness/determined/cli/trial.py (hunk 5 above)
Comment (MikhailKardash) [reply harness/determined/cli/trial.py:192]: I like that, updated to reflect that style.

File: harness/determined/cli/trial.py (hunk 4 above)
Comment (azhou-determined) [reply harness/determined/cli/trial.py:184]: but how would you even pick a different checkpoint UUID? we don't return a list of valid checkpoint UUIDs here, so it's not like we can tell users to call that and get the next best checkpoint UUID. this message might make sense if you specified `--uuid` directly, since that implies you already had some way of finding that UUID, but it's weird for `--best` since i'd assume the CLI would download the best valid checkpoint. 

i think that's what @wes-turner is trying to say. 

i agree that the ideal message will say something about what happened and give reasonable steps to resolve. here i think the best thing the user can do is call `det t ID describe` which will render a list of checkpoints for the trial, and from there, they can manually select the checkpoint they want to download. 

tbh, i would rather just delete this option entirely, and replace it with a `list-checkpoints` on `det trial`, which would just return a list of checkpoints with UUIDs, state, searcher metrics, etc. then the user can call `det c download UUID` for the one they want.

Comment (wes-turner) [reply harness/determined/cli/trial.py:184]: > tbh, i would rather just delete this option entirely, and replace it with a list-checkpoints on det trial, which would just return a list of checkpoints with UUIDs, state, searcher metrics, etc. then the user can call det c download UUID for the one they want.

Mmm. I like this strategy. As a long-term one, rather than for this ticket. Which is what I think you meant.

File: harness/determined/cli/trial.py (hunk 5 above)
Comment (wes-turner) [reply harness/determined/cli/trial.py:193]: Looking into `determined.cli.cli` here, it'd be nice if the raised exception could be localized there to exactly this problem. But I'm not sure it makes sense to try and hack that in. Probably what you decided, too.

I _do_ think there's some value in using `cli.die`. Value I see includes:
* Standardization -- when exiting via `die` we get pretty colors
* explicit "something when wrong"
* nonzero exit code

But it's a little weird that most of the dying is done in `cli.cli` and this would be the first(?) example of it elsewhere.

Have you thought about using `cli.die` here, yourself? What did you conclude?

Comment (MikhailKardash) [reply harness/determined/cli/trial.py:193]: Yeah, let's keep specific exceptions/error messages local to their respective CLI methods.

I also concluded that doing `cli.die` outside of `cli.cli` feels weird, but that definitely can change. I'd rather do standardizations like that in another larger and wide-reaching PR, especially given that this current PR is not about long-term strategy but rather a fix for a particular bug.

Comment (wes-turner) [reply harness/determined/cli/trial.py:193]: There's some weirdness for true. `cli.cli` thinks that it generally has all the necessary context to surface errors, but `cli.trial.download` thinks that it has context `cli.cli` doesn't. It's too bad if we can't satisfy `cli.cli` there in keeping with the existing standard, but I'm not going to lose too much sleep over it. I dunno. It _does_ feel like `cli.cli` is the right place to die rather than here.

No matter, I don't think keeping exceptions/error messages local to their respective CLI methods is the current standard.

If you don't want to import from `cli.cli` and you don't want to propagate this error to `cli.cli`, how do you feel about factoring `die` out of `cli.cli` and into `cli._util`? I think it's a slightly weird experience for users to see some errors printed in red and resulting in nonzero exit codes and other errors printed

Comment (wes-turner) [reply harness/determined/cli/trial.py:193]: As communicated OOB, I think it's important that an error here is handled the same way as other CLI errors are at the bottom of `cli.cli`. `download` _does_ have context here that `cli.cli` doesn't, but that context can be passed to `cli.cli` by catching and reraising the error.

File: harness/determined/cli/trial.py [hunk 6]
```diff
@@ -9,11 +9,12 @@
 from termcolor import colored
 
 from determined import cli
-from determined.cli import render
+from determined.cli import errors, render
 from determined.cli.master import format_log_entry
 from determined.common import api
 from determined.common.api import authentication, bindings
 from determined.common.declarative_argparse import Arg, ArgsDescription, Cmd, Group, string_to_bool
+from determined.common.experimental.checkpoint import CheckpointState
```
Comment (wes-turner) [inline harness/determined/cli/trial.py:17]: The SDK itself gets a pass, but everywhere else we should reference symbols from outside the `determined.common.experimental` (I think it already throws a warning if you import like this). `CheckpointState` is aliased into `client`.

To be forward-looking ([here's our doc](https://hpe.sharepoint.com/:w:/r/teams/detai/_layouts/15/Doc.aspx?sourcedoc=%7BCCA17AD5-E1DA-4660-AE14-97F22A58C1CB%7D&file=python%20code%20style.docx&action=default&mobileredirect=true), https://google.github.io/styleguide/pyguide.html#22-imports the Google style guide), please only import package / module names.

File: harness/determined/cli/trial.py [hunk 7]
```diff
@@ -175,11 +176,17 @@ def download(args: Namespace) -> None:
         checkpoints = det.get_trial(args.trial_id).list_checkpoints(
             sort_by=sort_by,
             order_by=order_by,
-            max_results=1,
         )
         if not checkpoints:
             raise ValueError(f"No checkpoints found for trial {args.trial_id}")
-        checkpoint = checkpoints[0]
+
+        valid_states = [CheckpointState.COMPLETED, CheckpointState.PARTIALLY_DELETED]
```
Comment (wes-turner) [inline harness/determined/cli/trial.py:183]: nit: "valid" is more ambiguous here than "downloadable", but because the definition is bound so tightly to its use it's not a big deal. Just a suggestion to adopt or not per your own preference.

Comment (wes-turner) [inline harness/determined/cli/trial.py:183]: So, something to think about... I don't love any decision, and I'll support whatever you decide:

It kind of sucks for `cli.trial.download` to have to know details about `Checkpoint.download` (namely, that `Checkpoint.download` only works for completed or partially deleted checkpoints). It kind of leaks implementation from the "checkpoint" abstraction.

A way around this is to just try/catch each `Checkpoint.download` / CheckpointStateException, no fewer details leaked.

But this isn't perfect because:
1. you still have to know that CheckpointStateExceptions are a thing
2. the implementation feels a little gross for reasons I can't quite verbalize

I think my own preference lies in this try/catch, but it's super easy for me to imagine people preferring the filter as written, and I think that's a fine choice, too.

If you also prefer it as written I support you in it. Just wanted to give you the prompt to think about it.

**Review by wes-turner — APPROVED** [review]
I like it!

File: harness/determined/cli/trial.py (hunk 7 above)
Comment (MikhailKardash) [reply harness/determined/cli/trial.py]: I think the filtering implementation is clearer. We don't have to obfuscate why `download` fails in this particular case.
